tests/unit/plugins/test_publisher.py

```diff
@@ -15,22 +15,409 @@
+
+    mocker.patch.object(http_client, 'request', post_request_coro)
+    mocker.patch.object(http_client, '_request_post_mock', request_post_mock)
+
+    return http_client
+
+
+@pytest.fixture
+def publisher_instance(mock_http_client, config):
+    success, error = asyncio.Queue(), asyncio.Queue()
+    pb = publisher.GDNSPublisher(config, success, error, mock_http_client)
+    return pb
+
+
+@pytest.fixture
+def mock_sleep(mocker):
+    sleep = mocker.Mock()
+
+    async def mock_sleep(*args, **kwargs):
+        sleep(*args, **kwargs)
+    mocker.patch('asyncio.sleep', mock_sleep)
+    return sleep
+
+
+#####
+# Publisher Tests
+#####
+def test_implements_interface(config, auth_client):
     """GDNSPublisher implements IPublisherClient"""
-    config = {'foo': 'bar'}
+    client = http.AIOConnection(auth_client=auth_client)
+
     success, error = asyncio.Queue(), asyncio.Queue()
-    client = publisher.GDNSPublisher(config, success, error)
+    plugin = publisher.GDNSPublisher(config, success, error, client)
 
-    assert interfaces.IPublisherClient.providedBy(client)
+    assert interfaces.IPublisherClient.providedBy(plugin)
     assert interfaces.IPublisherClient.implementedBy(publisher.GDNSPublisher)
-    assert config is client.config
-    assert success is client.success_channel
-    assert error is client.error_channel
-    assert 'publish' == client.phase
+    assert config is plugin.config
+    assert success is plugin.success_channel
+    assert error is plugin.error_channel
+    assert 'publish' == plugin.phase
+
+
+@pytest.mark.asyncio
+async def test_publish_changes_raises_exception_on_invalid_zone(
+        publisher_instance, event_message, event_msg_data_with_invalid_zone,
+        caplog):
+    """Ensure exception raised on invalid zone"""
+    event_message.data = event_msg_data_with_invalid_zone
+
+    with pytest.raises(exceptions.InvalidDNSZoneInMessageError) as error:
+        await publisher_instance.publish_changes(
+            event_message)
+    assert error.match('Error when asserting zone for record:')
+    assert 2 == len(caplog.records)
+
+
+@pytest.mark.asyncio
+async def test_publish_changes_handles_update_conflict(
+    publisher_instance, event_message,
+        all_existing_zone_records, rrsets_url, changes_url,
+        initial_changes_req, handled_conflict_changes_req,
+        initial_changes_pending_json_resp, initial_changes_resp, caplog):
+    """Ensure exception raised on invalid zone"""
+    expected_change_id = json.loads(initial_changes_pending_json_resp)['id']
+    event_message.data['resourceRecords'] = initial_changes_req['additions']
+    publisher_instance.http_client._request_post_mock.side_effect = [
+        exceptions.GCPHTTPError('409'), initial_changes_pending_json_resp]
+    publisher_instance.http_client._get_all_mock.return_value = (
+        all_existing_zone_records)
+    publisher_instance.http_client._get_json_mock.side_effect = [
```

Comment:
This works, but could be a return_value.

Reply:
ah good point - i think i originally had a few side effects